### PR TITLE
deps: fixing microtask behavior with multiple contexts

### PR DESCRIPTION
**Pull request by MSLaguana**
Initially we were handling microtasks per-context since our promise
interface relies on per-library callbacks, but the v8 API actually
deals with microtasks on a per-isolate level. This meant that we
weren't properly firing microtasks / promise resolutions from
contexts other than the currently active one.

Now we also handle the microtask queue at the isolate level.

Partially addresses #420 in that promises should now work in `vm.runInContext`

##### Checklist

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [X] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

### CODE REVIEW

**Comment by MSLaguana** [comment]
I'll add some tests for this in a bit, running the existing tests locally to make sure I didn't break anything obvious there.

**Comment by MSLaguana** [comment]
CI: https://ci.nodejs.org/job/chakracore-test-pull-request/259/

Turns out that one of node's upstream tests had been hit by this bug causing it not to run, and now that it does run it turns out it fails due to an error in upstream chakracore.

File: deps/chakrashim/src/jsrtisolateshim.cc [hunk 1]
```diff
@@ -670,6 +679,40 @@ void IsolateShim::SetPromiseRejectCallback(v8::PromiseRejectCallback callback) {
...
+        JsGetAndClearException(&notUsed);  // swallow any exception from task
+      }
+      JsRelease(task, nullptr);
+    }
+  }
+}
+
+void IsolateShim::QueueMicrotask(JsValueRef task) {
+  JsAddRef(task, nullptr);
+  this->microtaskQueue = new MicroTask(task, this->microtaskQueue);
```
Comment (digitalinfinity) [inline deps/chakrashim/src/jsrtisolateshim.cc]: What if this fails? Do we leak the task?

File: deps/chakrashim/src/jsrtisolateshim.cc [hunk 2]
```diff
@@ -670,6 +679,40 @@ void IsolateShim::SetPromiseRejectCallback(v8::PromiseRejectCallback callback) {
...
 }
 
+void IsolateShim::RunMicrotasks() {
+  // Loop until we've handled all the tasks, including the ones added by these tasks
+  while(this->microtaskQueue != nullptr) {
+    MicroTask* batch = this->microtaskQueue;
+    this->microtaskQueue = nullptr;
+
+    // Ensure that we handle things in-order
+    MicroTask* ordered = nullptr;
```
Comment (digitalinfinity) [inline deps/chakrashim/src/jsrtisolateshim.cc]: I don't follow- what is the intent here?

**Review by digitalinfinity — APPROVED** [review]

File: deps/chakrashim/src/jsrtisolateshim.cc (hunk 1 above)
Comment (MSLaguana) [reply deps/chakrashim/src/jsrtisolateshim.cc]: By "this" do you mean the `new` allocation? If that fails, then things are probably in a bad way (this is only allocating 2 pointers worth of data) but I hadn't given much thought to that scenario. The previous behavior was that we would push the task into an array in javascript, which I guess would OOM in the worst case and terminate the process, while here if the allocation fails then we leak any prior tasks, and also never run them.

File: deps/chakrashim/src/jsrtisolateshim.cc (hunk 2 above)
Comment (MSLaguana) [reply deps/chakrashim/src/jsrtisolateshim.cc]: Microtasks are supposed to run in the order that they are queued. For simplicity I've just implemented a stack, but to operate in FIFO order I just reverse the stack here.

Comment (digitalinfinity) [reply deps/chakrashim/src/jsrtisolateshim.cc]: Can we use a tail pointer and append to the end (i.e implement an actual queue)

I'd also just settle for a less terse comment

Comment (MSLaguana) [reply deps/chakrashim/src/jsrtisolateshim.cc]: Updated the comment

File: deps/chakrashim/src/jsrtisolateshim.cc [hunk 3]
```diff
@@ -670,6 +679,43 @@ void IsolateShim::SetPromiseRejectCallback(v8::PromiseRejectCallback callback) {
...
+    // in a FIFO manner 
+    MicroTask* ordered = nullptr;
+    for (MicroTask* t = batch; t != nullptr;) {
+      MicroTask* next = t->next;
+      t->next = ordered;
+      ordered = t;
+      t = next;
+    }
+
+    while (ordered != nullptr) {;
```
Comment (kfarnung) [inline deps/chakrashim/src/jsrtisolateshim.cc]: rogue semicolon?

File: deps/chakrashim/src/jsrtisolateshim.h [hunk 4]
```diff
@@ -164,6 +167,12 @@ class IsolateShim {
   void SetPromiseRejectCallback(v8::PromiseRejectCallback callback);
 
  private:
+  struct MicroTask {
+    MicroTask(JsValueRef task, MicroTask* next) : task(task), next(next) {}
```
Comment (kfarnung) [inline deps/chakrashim/src/jsrtisolateshim.h]: Could we have this manage the lifetime of the JsValueRef?

File: deps/chakrashim/src/jsrtisolateshim.cc [hunk 5]
```diff
@@ -670,6 +679,43 @@ void IsolateShim::SetPromiseRejectCallback(v8::PromiseRejectCallback callback) {
                                    reinterpret_cast<void*>(callback));
 }
 
+void IsolateShim::RunMicrotasks() {
+  // Loop until we've handled all the tasks, including the ones
+  // added by these tasks
+  while (this->microtaskQueue != nullptr) {
+    MicroTask* batch = this->microtaskQueue;
```
Comment (kfarnung) [inline deps/chakrashim/src/jsrtisolateshim.cc]: I'd recommend using std::vector and then using std::swap to get the batch:

```
std::vector<MicroTask> batch;
std::swap(this->microtaskQueue, batch);
```

**Comment by MSLaguana** [comment]
new CI: https://ci.nodejs.org/job/chakracore-test-pull-request/260/

File: deps/chakrashim/src/jsrtisolateshim.cc [hunk 6]
```diff
@@ -670,6 +671,26 @@ void IsolateShim::SetPromiseRejectCallback(v8::PromiseRejectCallback callback) {
...
+      JsValueRef notUsed;
+      if (jsrt::CallFunction(batch[i].task, &notUsed) != JsNoError) {
+        JsGetAndClearException(&notUsed);  // swallow any exception from task
+      }
+    }
+  }
+}
+
+void IsolateShim::QueueMicrotask(JsValueRef task) {
+  this->microtaskQueue.emplace_back(task);
```
Comment (MSLaguana) [inline deps/chakrashim/src/jsrtisolateshim.cc:691]: @kfarnung Would you mind double checking this is correct? I'm intending that this constructs the `MicroTask` in place in the vector

Comment (kfarnung) [reply deps/chakrashim/src/jsrtisolateshim.cc:691]: Looks correct to me.

File: deps/chakrashim/src/jsrtisolateshim.cc [hunk 7]
```diff
@@ -670,6 +671,26 @@ void IsolateShim::SetPromiseRejectCallback(v8::PromiseRejectCallback callback) {
...
 }
 
+void IsolateShim::RunMicrotasks() {
+  // Loop until we've handled all the tasks, including the ones
+  // added by these tasks
+  while (!this->microtaskQueue.empty()) {
+    std::vector<MicroTask> batch;
+    std::swap(batch, this->microtaskQueue);
+
+    for (unsigned int i = 0; i < batch.size(); ++i) {
```
Comment (kfarnung) [inline deps/chakrashim/src/jsrtisolateshim.cc]: You might be able to do:

`for (const MicroTask& microTask : batch) {`

which provides a const reference to the item in the vector (and thus doesn't copy it)

File: deps/chakrashim/src/jsrtisolateshim.cc [hunk 8]
```diff
@@ -670,6 +671,26 @@ void IsolateShim::SetPromiseRejectCallback(v8::PromiseRejectCallback callback) {
...
+      JsValueRef notUsed;
```
Comment (kfarnung) [inline deps/chakrashim/src/jsrtisolateshim.cc]: nit: Probably still worth initializing to `JS_INVALID_REFERENCE`

File: deps/chakrashim/src/jsrtisolateshim.h [hunk 9]
```diff
@@ -164,6 +167,24 @@ class IsolateShim {
   void SetPromiseRejectCallback(v8::PromiseRejectCallback callback);
 
  private:
+  struct MicroTask {
+    MicroTask(JsValueRef task) : task(task) {
+      JsAddRef(task, nullptr);
```
Comment (kfarnung) [inline deps/chakrashim/src/jsrtisolateshim.h]: Shouldn't matter, but maybe use `this->task` to disambiguate?

**Review by kfarnung — APPROVED** [review]
LGTM

**Comment by MSLaguana** [comment]
Third time's a charm: https://ci.nodejs.org/job/chakracore-test-pull-request/261/

**Pull request merged**